### PR TITLE
Issue #13: Sort item list by when to buy urgency

### LINKED ISSUE #13: 13. As a user, I want to view a list of my shopping list items in order of how soon I am likely to need to buy each of them again so that it’s clear what I need to buy soon
## Summary
Sorting items by how soon the user might need to buy them will make it easier for the user to identify what they need to buy, which will in turn help them better plan their shopping trips. To that end, we will be sorting items into 4 possible groups, according to how urgently the user should buy the item:

1. Need to buy soon (7 days or fewer until the next purchase)
2. Need to buy kind of soon (between 7 & 30 days until the next purchase)
3. Need to buy not soon (30 days or more until the next purchase)
4. Inactive (60 days have passed since the last purchase)

To implement this feature, your team will need to create some new UI elements so the user knows how their items are sorted and write some logic behind-the-scenes to perform the sorting itself. Read the acceptance criteria and our notes carefully to be sure you understand the tasks at hand, and don’t be afraid to reach out to the rest of your team, and your mentors if you need help. You’ve got this! 

## Acceptance criteria
- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [x] This urgency indicator *does not* rely on only color
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [x] sorts inactive items last, then
	- [x] sorts items in ascending order of days until purchase, and
	- [x] sorts items with the same days until purchase alphabetically

### A stretch goal
If you complete all of the previous acceptance criteria, consider what happens when an item’s `dateNextPurchased` has passed, but it isn’t yet inactive. Let’s call that item “overdue”.

- [x] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [x] Indicate in your UI when an item is overdue

## Notes
There's a lot to do for this issue! Read this section for some ideas about how to approach the tasks ahead. The `getDaysBetweenDates` function in `src/utils/dates.js` will be useful!

Remember: The dates in the Firestore database are special `Timestamp` objects. You’ll need to call their `toDate()` method to get a JavaScript Date that we can work with.

### Indicating purchase urgency in the UI
It’s tempting to use colors or icons to indicate how soon the user should buy an item. Keep in mind:
- If you use only color, a colorblind user won’t be able to tell the difference between urgency indicators
- If you use icons, you have to teach users what those icons mean

With these constraints, it’s probably best to render plain text, like “soon”, “kind of soon”, etc. in order to show urgency. You can still use color alongside that text, and the design of where that text goes is up to you!

### Sorting the list
We’re sorting by multiple criteria:
1. Whether the item is active or inactive, then
2. By the number of days until its `dateNextPurchased`, and
3. Alphabetically by name, if they have the same number of days until purchase

Sorting by multiple criteria can be tricky! Refer to the MDN docs for [`Array.prototype.sort`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) and refresh yourself on how to write *compare functions* so you can write the `comparePurchaseUrgency` function.

### PR DESCRIPTION
**Pull request by bbland1**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._

## Description

Adds a text indicator of the buying urgency status of an item to the right of the item and sorts the list based on the status. If an item is inactive it will be at the bottom of the list, overdue items will be at the top, and if any items have the same days until next purchase they will be additionally sorted alphabetically.

## Related Issue

Closes #13 

## Acceptance Criteria

- [x] Items in the list are shown with an indicator that tells the user they should buy the item “soon”, “kind of soon”, or “not soon”; or that the item is “inactive”
	- [x] This urgency indicator *does not* rely on only color
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors
	- [x] sorts inactive items last, then
	- [x] sorts items in ascending order of days until purchase, and
	- [x] sorts items with the same days until purchase alphabetically
- [x] Extend the functionality of `comparePurchaseUrgency` to sort “overdue” items to the top of the list
- [x] Indicate in your UI when an item is overdue

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|  ✓    | 💯 Enhancement |
|  ✓    | ✨ New Feature |

### Before

Screenshot 2024-09-19 at 12 21 14 AM

### After

Screenshot 2024-09-19 at 12 21 39 AM

## Testing Steps / QA Criteria

- Click the [preview link](https://tcl-77-smart-shopping-list--pr35-bb-fz-sort-by-purcha-99ypueis.web.app/) for this PR
- Sign in with your account
- Click a list with item and navigate using the nav bar
- Items already on list should be sorted based on urgency status
- To see other status messages go to the [firebase](https://console.firebase.google.com/u/2/project/tcl-77-smart-shopping-list/firestore/data/~2F)
- Open the list and item you want to change
- Change to various dates to see status change

### CODE REVIEW

File: src/api/README.md [hunk 1]
```diff
@@ -55,15 +55,24 @@ This function takes user-provided data and uses it to update an exiting item in
 | Parameter  | Type     | Description                                                     |
 | ---------- | -------- | --------------------------------------------------------------- |
 | `listPath` | `string` | The Firestore path of the list to which the item will be added. |
-| `item`     | `string` | The item object.                                                |
+| `item`     | `object` | The item object.                                                |
```
Comment (tannaurus) [inline src/api/README.md:58]: I love that we're making updates here! I might suggest that we change `item` from just `object` to include the fields that we are anticipating on that object. You can do that like this:

```
| `item.name`     | `string` | The name of your item                                             |
(ect etc)
```

File: src/api/README.md [hunk 2]
```diff
@@ -46,7 +46,7 @@ This function takes user-provided data and uses it to create a new item in the F
 
 #### Note
 
-**`daysUntilNextPurchase` is not added to the item directly**. It is used alomngside the `getFutureDate` utility function to create a new _JavaScript Date_ that represents when we think the user will buy the item again.
+**`daysUntilNextPurchase` is not added to the item directly**. It is used alongside the `getFutureDate` utility function to create a new _JavaScript Date_ that represents when we think the user will buy the item again.
```
Comment (tannaurus) [inline src/api/README.md:49]: nice catch :) I live and die by spell check in my ide for this reason. I shipped one too many typos to production, I swore I'd never do it again (so far so good 🚀 )

File: src/api/firebase.ts [hunk 3]
```diff
@@ -301,3 +301,44 @@ export async function deleteItem() {
...
+	const currentDate = new Date();
+
+	// Check for last time item had any activity
+	const daysSinceItemLastActivity = item1.dateLastPurchased
+		? getDaysBetweenDates(currentDate, item1.dateLastPurchased.toDate())
+		: getDaysBetweenDates(currentDate, item1.dateCreated.toDate());
+
+	// Prioritize item1 if current date is past next purchase date
+	if (currentDate > item1.dateNextPurchased.toDate()) {
+		return -1;
```
Comment (tannaurus) [inline src/api/firebase.ts:315]: I really like this! Did you consider an enum here? There's pros and cons:
- Enums are much easier to read/understand (and therefore produce more maintainable* code). You could have an enum that contains three variants: `Priority.Low`, `Priority.Medium`, `Priority.High`. Now you don't need to write down what `-1` means. To some, that might be misunderstood to be _very_ low priority. Careful documentation can get around that. You can also assign enums values ([see some of the examples here](https://www.typescriptlang.org/docs/handbook/enums.html)) so you can still keep the numbers where you need them (if we're putting this in a database, for instance) but give them "more meaning" in the code.
- Numbers along are really great if you plan on adding more priorities in the future, which isn't _really_ a thing here, but something to keep in mind. This is something I've seen done with errors before, where a function will have a big switch statement against every error variant a function could return, and assign a numerical value to the significance of that error. Assuming you want to trust someone else's evaluation of "significance" there, I don't think I would in production code but i'm just paranoid, you could decide to handle the "top 30% of significant errors" differently than "lower priority" errors. Imo this is typically the result of less-than-great api design and could be avoided in other ways, but I'll try to avoid a soap box here 😄 

tldr; maybe consider enums with values assigned to them 🚀

Comment (tannaurus) [reply src/api/firebase.ts:315]: 🤦  I just realized this is for a `sort` - forget everything I said! Maybe we could leave a comment above this function explaining what it's used for and a brief overview of what it considers a "high priority" vs "low priority" ?

File: src/components/ListItem.tsx [hunk 4]
```diff
@@ -29,6 +29,36 @@ export function ListItemCheckBox({ item, listPath }: Props) {
...
+
+		const daysSinceItemLastActivity = item.dateLastPurchased
+			? getDaysBetweenDates(currentDate, item.dateLastPurchased.toDate())
+			: getDaysBetweenDates(currentDate, item.dateCreated.toDate());
+
+		if (currentDate > item.dateNextPurchased.toDate()) {
+			return "overdue";
+		}
+
+		if (daysSinceItemLastActivity >= 60) {
```
Comment (tannaurus) [inline src/components/ListItem.tsx:47]: These are great examples of guard clauses!

File: src/components/ListItem.tsx [hunk 5]
```diff
@@ -29,6 +29,36 @@ export function ListItemCheckBox({ item, listPath }: Props) {
...
+		const daysUntilNextPurchase = getDaysBetweenDates(
+			new Date(),
+			item.dateNextPurchased.toDate(),
+		);
+
+		const daysSinceItemLastActivity = item.dateLastPurchased
+			? getDaysBetweenDates(currentDate, item.dateLastPurchased.toDate())
+			: getDaysBetweenDates(currentDate, item.dateCreated.toDate());
+
+		if (currentDate > item.dateNextPurchased.toDate()) {
```
Comment (tannaurus) [inline src/components/ListItem.tsx:43]: _tiny_ suggestion that's damn near a nit pick

something I encourage is thinking a lot about where you initialize variables/perform logic inside "groupings" in logically a reasonable order. This function violates that a bit: `daysUntilNextPurchase` is calculated at the top, but it isn't' used until line `51` - this could mean that we're doing an operation that _might_ have some performance cost that we never needed to "spend" (I don't think it's a big deal here, just something to consider generally).

Things like this also make code a bit harder to read. When reviewing logic, it's convenient to glance up a few lines (when possible) and see the value calculated that's then being compared in `if`s

`daysSinceItemLastActivity` is another example of this. It might not ever be used if line 44 returns

**Review by tannaurus — APPROVED** [review]

File: src/api/firebase.ts [hunk 6]
```diff
@@ -301,3 +301,44 @@ export async function deleteItem() {
 	 * this function must accept!
 	 */
 }
+
+export function comparePurchaseUrgency(item1: ListItem, item2: ListItem) {
```
Comment (tannaurus) [inline src/api/firebase.ts:305]: You might be able to update this function's return type to be a bit more explicit

```
export function comparePurchaseUrgency(item1: ListItem, item2: ListItem): -1 | 0 | 1 {
```

File: src/api/README.md (hunk 2 above)
Comment (bbland1) [reply src/api/README.md:49]: yeah I learned long ago just how bad I can be at spelling 😅 so I looked for a [spell checker extension](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) in VS to support me when I fail 😂

File: src/views/authenticated/List.tsx [hunk 7]
```diff
@@ -14,9 +14,11 @@ export function List({ data: unfilteredListItems, listPath }: Props) {
...
 
 	const filteredListItems = useMemo(() => {
-		return unfilteredListItems.filter((item) =>
-			item.name.toLowerCase().includes(searchTerm.toLowerCase()),
-		);
+		return unfilteredListItems
+			.filter((item) =>
+				item.name.toLowerCase().includes(searchTerm.toLowerCase()),
+			)
+			.sort(comparePurchaseUrgency);
```
Comment (RossaMania) [inline src/views/authenticated/List.tsx:21]: PRAISE: I like the use of .sort() here.

**Review by RossaMania — APPROVED** [review]
I like this code.

Everything seems straightforward, and tied together nicely with that .sort() method!

File: src/api/firebase.ts [hunk 8]
```diff
@@ -301,3 +301,44 @@ export async function deleteItem() {
...
+		getDaysBetweenDates(currentDate, item1.dateNextPurchased.toDate()) >
+		getDaysBetweenDates(currentDate, item2.dateNextPurchased.toDate())
+	) {
+		return 1;
+		//if both items have the same sort order, we sort alphabetically
+	} else {
+		if (item1.name.toLowerCase() < item2.name.toLowerCase()) {
+			return -1;
+		} else {
+			return 1;
```
Comment (alex-andria) [inline src/api/firebase.ts:341]: Nice work figuring out this logic!🌟🌟 It's a real brain twister at the start but this is very well done. know in the past others have explored switch functions for a cleaner look and can be faster to compile depending on the use case. Like DJ said above this could also just be tightened up with turning some `else if` statements to `if` statements too.

File: src/components/ListItem.tsx [hunk 9]
```diff
@@ -29,6 +29,36 @@ export function ListItemCheckBox({ item, listPath }: Props) {
...
+		if (daysSinceItemLastActivity >= 60) {
+			return "inactive";
+		}
+
+		if (daysUntilNextPurchase >= 30) {
+			return "not soon";
+		}
+
+		if (daysUntilNextPurchase <= 7) {
+			return "buy soon";
```
[the remark this answers is not in the document]
Comment (alex-andria) [reply src/components/ListItem.tsx:56]: +1 on ommiting buy

**Review by alex-andria — APPROVED** [review]
Works great in testing it out and checking firebase, great work Falak and Brianna! 🙌 🙌 Added my comments but you knocked the logic for this out the park as I know this is a tricky one myself!

File: src/views/authenticated/List.tsx (hunk 7 above)
Comment (eternalmaha) [reply src/views/authenticated/List.tsx:21]: I agree! Great detail to implement the sorting even within the filtered list.

**Comment by eternalmaha** [comment]
Wow!! Incredible work on the logic for this issue. When I was playing around with it, it was interesting to see the logic play out if I "purchased" an existing or new item earlier than expected (it immediately registering it as "overdue").  My suggestion for UI purposes, perhaps we could use a different word other than "overdue". I feel like it may confuse the user who might think they need to always buy the item immediately ? The other messages indicate a priority message ("soon", "kind of soon", "not soon"). Maybe along those lines, it could be "very soon" instead of "overdue"? My humble opinion!

**Review by eternalmaha — APPROVED** [review]

File: src/views/authenticated/List.tsx (hunk 7 above)
Comment (eternalmaha) [reply src/views/authenticated/List.tsx:21]: I gotta pick your guys' brain a bit tomorrow about the logic of this piece of code :)
